tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <memory>
+#include <type_traits>
+#include <version>
+
+#ifdef __cpp_lib_optional
+#include <optional>
+#endif
+
+#ifdef __cpp_lib_any
+#include <any>
+#endif
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+using namespace std;
+
+// tests for shared_ptr<T>::operator=
+template <class T, class U, class = void>
+constexpr bool can_shared_ptr_assign = false;
+
+template <class T, class U>
+constexpr bool can_shared_ptr_assign<T, U, void_t<decltype(declval<shared_ptr<T>&>() = declval<U>())>> = true;
+
+STATIC_ASSERT(!can_shared_ptr_assign<int, shared_ptr<long>>);
+STATIC_ASSERT(!can_shared_ptr_assign<int, const shared_ptr<long>&>);
+STATIC_ASSERT(!can_shared_ptr_assign<int, unique_ptr<long>>);
+#if _HAS_AUTO_PTR_ETC
+STATIC_ASSERT(!can_shared_ptr_assign<int, auto_ptr<long>>);
+#endif

```

Comment:
This is good, but I recommend having one `STATIC_ASSERT` where the handwritten type trait is `true` - this verifies that the type trait works (and isn't always `false` for some other reason).

Occurs below.